### PR TITLE
fix: index out of bound when chunking

### PR DESCRIPTION
**Pull request by nicolas-vivot**
This PR fixes a certain number of issues when using the iterator from `NEChunks`.

I also added / enriched tests to ensure the missed edge cases are not going to blow up on next modification.

### CODE REVIEW

File: src/slice.rs [hunk 1]
```diff
@@ -143,35 +144,39 @@ impl<'a, T> NonEmptyIterator for NEChunks<'a, T> {
 
     fn next(&mut self) -> Option<Self::Item> {
         if self.index == 0 {
-            let end = self.window.get() - 1;
+            let end = min(self.window.get() - 1, self.tail.len());
```
Comment (nicolas-vivot) [inline src/slice.rs:147]: This is required to ensure cases where we have less elements in the collection than the given window.

File: src/slice.rs [hunk 2]
```diff
@@ -143,35 +144,39 @@ impl<'a, T> NonEmptyIterator for NEChunks<'a, T> {
...
             let slice = NESlice {
                 head: self.head,
                 tail: &self.tail[0..end],
             };
 
-            self.index = end;
+            // 2024-03-18: This is a workaround for edge case of 1 element slice, thus the index
+            // does not get incremented as end is 0, thus the next iteration will return the same
+            // slice gain and again.
+            self.index = end + 1;
```
Comment (nicolas-vivot) [inline src/slice.rs:157]: Since we do not allow to have the index starting at -1, this is a trick to avoid endless looping on the iterator when we have a single element on the collection.
Obviously this has an impact on the comparison with the tail length for the stop case.

File: src/slice.rs [hunk 3]
```diff
@@ -143,35 +144,39 @@ impl<'a, T> NonEmptyIterator for NEChunks<'a, T> {
...
 
             Some(slice)
-        } else if self.index >= self.tail.len() {
+        } else if self.index >= (self.tail.len() + 1) {
             None
         } else {
-            let end = self.index + self.window.get();
-            let slc: &'a [T] = &self.tail[self.index..end];
+            // Ensure we never go out of bounds
+            let end = min(self.index - 1 + self.window.get(), self.tail.len());
```
Comment (nicolas-vivot) [inline src/slice.rs:164]: Again, ensuring we do not go out of bounds when approaching the tail.

File: src/slice.rs [hunk 4]
```diff
@@ -143,35 +144,39 @@ impl<'a, T> NonEmptyIterator for NEChunks<'a, T> {
...
+                    self.index = end + 1;
                     Some(slice)
                 }
             }
         }
     }
 
     fn first(self) -> (Self::Item, Self::IntoIter) {
-        let end = self.window.get() - 1;
+        let end = min(self.window.get() - 1, self.tail.len());
```
Comment (nicolas-vivot) [inline src/slice.rs:179]: Same fix as for `next`.

File: src/slice.rs [hunk 5]
```diff
@@ -143,35 +144,39 @@ impl<'a, T> NonEmptyIterator for NEChunks<'a, T> {
...
+            // Ensure we never go out of bounds
+            let end = min(self.index - 1 + self.window.get(), self.tail.len());
+            let slc: &'a [T] = &self.tail[self.index - 1..end];
 
             match slc {
                 [] => None,
                 [head, tail @ ..] => {
                     let slice = NESlice { head, tail };
-                    self.index = end;
+                    self.index = end + 1;
```
Comment (nicolas-vivot) [inline src/slice.rs:171]: Staying consistent by keeping the index at +1

File: src/slice.rs [hunk 6]
```diff
@@ -143,35 +144,39 @@ impl<'a, T> NonEmptyIterator for NEChunks<'a, T> {
...
             Some(slice)
-        } else if self.index >= self.tail.len() {
+        } else if self.index >= (self.tail.len() + 1) {
             None
         } else {
-            let end = self.index + self.window.get();
-            let slc: &'a [T] = &self.tail[self.index..end];
+            // Ensure we never go out of bounds
+            let end = min(self.index - 1 + self.window.get(), self.tail.len());
+            let slc: &'a [T] = &self.tail[self.index - 1..end];
```
Comment (nicolas-vivot) [inline src/slice.rs:165]: Lowering the index by 1 since we had to increase it by 1 to cover the edge case mentioned above.

File: src/slice.rs [hunk 7]
```diff
@@ -157,7 +157,8 @@ impl<'a, T> NonEmptyIterator for NEChunks<'a, T> {
         } else if self.index >= self.tail.len() {
             None
         } else {
-            let end = self.index + self.window.get();
+            // Ensure we never go out of bounds
+            let end = min(self.index + self.window.get(), self.tail.len());
```
Comment (fosskers) [inline src/slice.rs]: Number types already have a `min` method provided by the `Ord` trait, so you don't need to important an extra function just for that.

File: src/slice.rs [hunk 8]
```diff
@@ -157,7 +157,8 @@ impl<'a, T> NonEmptyIterator for NEChunks<'a, T> {
...
             let slc: &'a [T] = &self.tail[self.index..end];
```
Comment (fosskers) [inline src/slice.rs]: My incorrect assumption was based on this:
```rust
fn main() {
    let v = vec![1, 2, 3];
    let u = &v[0..];

    println!("{:?}", u);
}
```
This does not panic. But this does:
```rust
fn main() {
    let v = vec![1, 2, 3];
    let u = &v[0..10];

    println!("{:?}", u);
}
```

File: src/slice.rs [hunk 9]
```diff
@@ -308,4 +314,28 @@ mod tests {
...
+        let v = nev![1, 2, 3, 4, 5, 6, 57];
+        let n = NonZeroUsize::new(3).unwrap();
+        let c: crate::slice::NEChunks<'_, i32> = v.nonempty_chunks(n);
+
+        let mut r: Vec<NESlice<i32>> = vec![];
+
+        for slice in c {
+            let _: NESlice<'_, i32> = slice;
+            r.push(slice);
+        }
```
Comment (fosskers) [inline src/slice.rs:333]: You can also just `collect` instead of manually pushing into a vector.

**Comment by nicolas-vivot** [comment]
LGTM
